tests/functional/directives.ts

```diff
@@ -0,0 +1,655 @@
+      }
+
+      @ObjectType()
+      class SampleObjectType {
+        @Field()
+        @Directive("foo")
+        withDirective: string = "withDirective";
+
+        @Field()
+        @Directive("bar", { baz: "true" })
+        withDirectiveWithArgs: string = "withDirectiveWithArgs";
+
+        @Field()
+        @Directive("upper")
+        withUpper: string = "withUpper";
+
+        @Field()
+        @Directive("@upper")
+        withUpperSDL: string = "withUpperSDL";
+
+        @Field()
+        @Directive("append")
+        withAppend: string = "hello";
+
+        @Field()
+        @Directive("@append")
+        withAppendSDL: string = "hello";
+
+        @Field()
+        @Directive("append")
+        @Directive("upper")
+        withUpperAndAppend: string = "hello";
+
+        @Field()
+        @Directive("@append @upper")
+        withUpperAndAppendSDL: string = "hello";
+
+        @Field()
+        withInput(@Arg("input") input: DirectiveOnFieldInput): string {
+          return `hello${input.append}`;
+        }
+
+        @Field()
+        @Directive("upper")
+        withInputUpper(@Arg("input") input: DirectiveOnFieldInput): string {
+          return `hello${input.append}`;
+        }
+
+        @Field()
+        withInputOnClass(@Arg("input") input: DirectiveOnClassInput): string {
+          return `hello${input.append}`;
+        }
+
+        @Field()
+        @Directive("upper")
+        withInputUpperOnClass(@Arg("input") input: DirectiveOnClassInput): string {
+          return `hello${input.append}`;
+        }
+      }
+
+      @Resolver()
+      class SampleResolver {
+        @Query(() => SampleObjectType)
+        objectType(): SampleObjectType {
+          return new SampleObjectType();
+        }
+
+        @Query()
+        @Directive("foo")
+        queryWithDirective(): string {
+          return "queryWithDirective";
+        }
+
+        @Query()
+        @Directive("bar", { baz: "true" })
+        queryWithDirectiveWithArgs(): string {
+          return "queryWithDirectiveWithArgs";
+        }
+
+        @Query()
+        @Directive("upper")
+        queryWithUpper(): string {
+          return "queryWithUpper";
+        }
+
+        @Query()
+        @Directive("@upper")
+        queryWithUpperSDL(): string {
+          return "queryWithUpper";
+        }
+
+        @Query()
+        @Directive("append")
+        queryWithAppend(): string {
+          return "hello";
+        }
+
+        @Query()
+        @Directive("@append")
+        queryWithAppendSDL(): string {
+          return "hello";
+        }
+
+        @Query()
+        @Directive("append")
+        @Directive("upper")
+        queryWithUpperAndAppend(): string {
+          return "hello";
+        }
+
+        @Query()
+        @Directive("@append @upper")
+        queryWithUpperAndAppendSDL(): string {
+          return "hello";
+        }
+
+        @Mutation()
+        @Directive("foo")
+        mutationWithDirective(): string {
```

Comment:
mutation and queries are handled 1:1 the same, there's no need for additional tests for them

Reply:
I initially was building the metadata in it's own build call after everything else was built, so I had to make sure queries and mutations were caught.  But I since then moved it back to building within each type's metadata build steps.  I can remove mutation tests, but if it for some reason gets moved then things'll break.

Reply:
I don't tests many things that works for mutation too because they are 1:1 the same as query, so it's not a problem to remove the duplicated tests. They are a maintenance problem as we have to fix tests twice because of that.